src/qt/walletcontroller.cpp

```diff
@@ -147,23 +146,64 @@ void WalletController::removeAndDeleteWallet(WalletModel* wallet_model)
     delete wallet_model;
 }
 
+WalletControllerActivity::WalletControllerActivity(WalletController* wallet_controller, QWidget* parent_widget)
+    : QObject(wallet_controller)
+    , m_wallet_controller(wallet_controller)
+    , m_parent_widget(parent_widget)
+{
+}
 
-OpenWalletActivity::OpenWalletActivity(WalletController* wallet_controller, const std::string& name)
-    : m_wallet_controller(wallet_controller)
-    , m_name(name)
-{}
+WalletControllerActivity::~WalletControllerActivity()
+{
+    delete m_progress_dialog;
+}
 
-void OpenWalletActivity::open()
+void WalletControllerActivity::showProgressDialog(const QString& label_text)
 {
-    std::string error, warning;
-    std::unique_ptr<interfaces::Wallet> wallet = m_wallet_controller->m_node.loadWallet(m_name, error, warning);
-    if (!warning.empty()) {
-        Q_EMIT message(QMessageBox::Warning, QString::fromStdString(warning));
-    }
-    if (wallet) {
-        Q_EMIT opened(m_wallet_controller->getOrCreateWallet(std::move(wallet)));
-    } else {
-        Q_EMIT message(QMessageBox::Critical, QString::fromStdString(error));
+    m_progress_dialog = new QProgressDialog(m_parent_widget);
+
+    m_progress_dialog->setLabelText(label_text);
+    m_progress_dialog->setRange(0, 0);
+    m_progress_dialog->setCancelButton(nullptr);
+    m_progress_dialog->setWindowModality(Qt::ApplicationModal);
+    GUIUtil::PolishProgressDialog(m_progress_dialog);
+
+    connect(m_progress_dialog, &QObject::destroyed, [this] {
+        m_progress_dialog = nullptr;
+    });
+}
+
+OpenWalletActivity::OpenWalletActivity(WalletController* wallet_controller, QWidget* parent_widget)
+    : WalletControllerActivity(wallet_controller, parent_widget)
+{
+}
+
+void OpenWalletActivity::finish()
+{
+    m_progress_dialog->hide();
+
+    if (!m_error_message.empty()) {
+        QMessageBox::critical(m_parent_widget, tr("Open wallet failed"), QString::fromStdString(m_error_message));
+    } else if (!m_warning_message.empty()) {
+        QMessageBox::warning(m_parent_widget, tr("Open wallet warning"), QString::fromStdString(m_warning_message));
     }
+
+    if (m_wallet_model) Q_EMIT opened(m_wallet_model);
+
     Q_EMIT finished();
 }
+
+void OpenWalletActivity::open(const std::string& path)
+{
+    QString name = path.empty() ? QString("["+tr("default wallet")+"]") : QString::fromStdString(path);
+
+    showProgressDialog(tr("Opening Wallet <b>%1</b>...").arg(name.toHtmlEscaped()));
+
+    QTimer::singleShot(500, worker(), [this, path] {
+        std::unique_ptr<interfaces::Wallet> wallet = node().loadWallet(path, m_error_message, m_warning_message);
+
+        if (wallet) m_wallet_model = m_wallet_controller->getOrCreateWallet(std::move(wallet));
+
+        QTimer::singleShot(500, this, &OpenWalletActivity::finish);
```

Comment:
Same here.

Reply:
`OpenWalletActivity::finish()` could be inlined here.
`OpenWalletActivity::open()` will remain small enough and readability will be improved, no?

Reply:
> OpenWalletActivity::finish() could be inlined here.

At the time I had 2 reasons to not do that:
 1. `finish` could be called from multiple places, also initially it was a virtual method in the super class;
 2. wanted to avoid nesting lambdas.

Reply:
Could it be just
```c++
        finish();
```
?

Reply:
No, this `finish()` is not thread safe - it would be called from the worker thread.

Reply:
>  ... this `finish()` is not thread safe - it would be called from the worker thread.

Right. But thread safety is not required for `finish()`, IMO.

Reply:
But it is, see `OpenWalletActivity::finish()`, it must be called on the GUI thread.

Reply:
Correct.